File: src/2022-intern/1462.course-schedule-iv.js
/*
 * @lc app=leetcode id=1462 lang=javascript
 *
 * [1462] Course Schedule IV
 */

// @lc code=start
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @param {number[][]} queries
 * @return {boolean[]}
 */
const checkIfPrerequisite = function (numCourses, prerequisites, queries) {
  /** @type {boolean[][]} */
  const courseMatrix = new Array(numCourses);
  for (let i = 0; i < numCourses; i++) {
    courseMatrix[i] = new Array(numCourses);
    courseMatrix[i].fill(false);
  }
  for (const [course1, course2] of prerequisites) {
    courseMatrix[course1][course2] = true;
  }

  let changed = true;
  while (changed) {
    changed = false;
    for (let i = 0; i < numCourses; i++) {
      for (let j = 0; j < numCourses; j++) {
        if (courseMatrix[i][j]) {
          for (let k = 0; k < numCourses; k++) {
            if (courseMatrix[j][k]) {
              if (!courseMatrix[i][k]) {
                changed = true;
              }
              courseMatrix[i][k] = true;
            }
          }
        }
      }
    }
  }

  /** @type {boolean[]} */
  const result = [];
  for (const [course1, course2] of queries) {
    result.push(courseMatrix[course1][course2]);
  }
  return result;
};
// @lc code=end

File: src/2025/1462.course-schedule-iv.cpp
/*
 * @lc app=leetcode id=1462 lang=cpp
 *
 * [1462] Course Schedule IV
 */

#include <vector>

// @lc code=start
class Solution {
 public:
  std::vector<bool> checkIfPrerequisite(
      const int numCourses,
      const std::vector<std::vector<int>>& prerequisites,
      const std::vector<std::vector<int>>& queries) {
    std::vector<std::vector<bool>> isReachable(
        numCourses, std::vector<bool>(numCourses, false));
    for (const auto& prerequisite : prerequisites) {
      const int from = prerequisite[0];
      const int to = prerequisite[1];
      isReachable[from][to] = true;
    }

    for (int m = 0; m < numCourses; m++) {
      for (int i = 0; i < numCourses; i++) {
        for (int j = 0; j < numCourses; j++) {
          isReachable[i][j] =
              isReachable[i][j] || (isReachable[i][m] && isReachable[m][j]);
        }
      }
    }

    std::vector<bool> queryResults(queries.size(), false);
    for (int i = 0; i < queries.size(); i++) {
      queryResults[i] = isReachable[queries[i][0]][queries[i][1]];
    }

    return queryResults;
  }
};
// @lc code=end
